src/rgw/rgw_pubsub.cc

```diff
@@ -24,19 +24,19 @@ void set_event_id(std::string& id, const std::string& hash, const utime_t& ts) {
 
 void rgw_s3_key_filter::dump(Formatter *f) const {
   if (!prefix_rule.empty()) {
-    f->open_object_section("FilterRule");
+    f->open_object_section("PrefixFilterRule");
```

Comment:
according to this: https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketNotificationConfiguration.html
the tags should be just "FilterRule", and the type fo teh rule "prefix/suffix/regex" should be in the "Name"

Reply:
thats what i wrote in the commit message,  that rule is for XML response. this is json response and is used by radosgw-admin which makes the key invalid as all the keys have the same name.i have not changed the dump_xml, it will still return FilterRule

Reply:
aws cli also return json. i think what they are doing is returing that as a list of objects, so that the same object can reepeat . we should probabloy fix the output to match that.
(see: https://docs.aws.amazon.com/cli/latest/reference/s3api/get-bucket-notification-configuration.html).

Reply:
> aws cli also return json. i think what they are doing is returing that as a list of objects, so that the same object can reepeat . we should probabloy fix the output to match that. (see: https://docs.aws.amazon.com/cli/latest/reference/s3api/get-bucket-notification-configuration.html).

so the output will look like
`S3Key": {
                    "FilterRule": [{
                        "Name": "prefix",
                        "Value": "health"
                    },
                   {
                        "Name": "suffix",
                        "Value": "old"
                    }]
                },`

Reply:
Converted to array, same as when you send the data via boto.the output will now be ```        "FilterRule": [
            {
                "Name": "prefix",
                "Value": "world"
            },
            {
                "Name": "suffix",
                "Value": "log"
            }
        ],```

Reply:
i ran the actual aws cli. setting the notification (after bucket and topic created):
```
aws --region=default --endpoint-url http://localhost:8000 s3api put-bucket-notification-configuration \                      
   --bucket fishbucket --notification-configuration='{"TopicConfigurations": [{"Id": "notif1", "TopicArn": "arn:aws:sns:default::fishtopic", "Events": [],  "Filter": {"Metadata": {"FilterRules": [{"Name": "hello", "Value": "world"}]}, "Tags": {"FilterRules": [{"Name": "hello", "Value": "world"}]}, "Key": {"FilterRules": [{"Name": "prefix", "Value": "world"}, {"Name": "suffix", "Value": "log"}]}}}]}'
```

when fetching the notification usin the aws cli:
```
aws --region=default --endpoint-url http://localhost:8000 s3api get-bucket-notification-configuration \
  --bucket fishbucket
```
I got the following json output (they convert our XML reply to json):
```json
{
  "TopicConfigurations": [
        {
            "Id": "notif1",
            "TopicArn": "arn:aws:sns:default::fishtopic",
            "Events": [
                "s3:ObjectCreated:*",
                "s3:ObjectRemoved:*"
            ],
            "Filter": {
                "Key": {
                    "FilterRules": [
                        {
                            "Name": "prefix",
                            "Value": "world"
                        },
                        {                           
                            "Name": "suffix",
                            "Value": "log"
                        }
                    ]
                },
                "Metadata": {
                    "FilterRules": [
                        {
                            "Name": "hello",
                            "Value": "world"
                        }
                    ]
                },
                "Tags": {
                    "FilterRules": [
                        {
                            "Name": "hello",
                            "Value": "world"
                        }
                    ]
                }
            }
        }
    ]
}

```

Reply:
based on that can you please also:
* change  the values in: `rgw_s3_filter::dump()` to: 'Key`, `Metadata` and `Tags`
* apply the same fix to: ` rgw_s3_key_value_filter::dump()` so we get the correct json format for tags and metadata filters

maybe better to split to another tracker/commit/pr?
so we can more easily backport the fix.

Reply:
Good catch.
added fix for both, created the tracker and also updated the commit message to reflect all 3 fixes